src/marqo/tensor_search/delete_docs.py

```diff
@@ -0,0 +1,96 @@
+"""
+This module handles the delete documents endpoint
+"""
+import datetime
+from typing import List, NamedTuple, Literal
+import json
+from marqo import errors
+from marqo._httprequests import HttpRequests
+from marqo.config import Config
+from marqo.tensor_search import validation, utils, enums
+
+
+# -- Marqo delete endpoint interface: --
+
+
+class MqDeleteDocsResponse(NamedTuple):
+    """An object that holds the data we send back to users"""
+    index_name: str
+    status_string: Literal["succeeded"]
+    document_ids: List[str]
+    deleted_docments_count: int
+    deletion_start: datetime.datetime
+    deletion_end: datetime.datetime
+
+
+def format_delete_docs_response(marqo_response: MqDeleteDocsResponse) -> dict:
+    """This formats the delete response for users """
+    return {
+        "index_name": marqo_response.index_name, "status": marqo_response.status_string,
+        "type": "documentDeletion", "details": {
+            "receivedDocumentIds": len(marqo_response.document_ids),
+            "deletedDocuments": marqo_response.deleted_docments_count,
+        },
+        "duration": utils.create_duration_string(marqo_response.deletion_end - marqo_response.deletion_start),
+        "startedAt": utils.format_timestamp(marqo_response.deletion_start),
+        "finishedAt": utils.format_timestamp(marqo_response.deletion_end),
+    }
+
+
+class MqDeleteDocsRequest(NamedTuple):
+    """An object that holds the data from users for a delete request"""
+    index_name: str
+    document_ids: List[str]
+    auto_refresh: bool
+
+
+# -- Data-layer agnostic logic --
+
+
+def delete_documents(config: Config, del_request: MqDeleteDocsRequest) -> dict:
+    """entrypoint function for deleting documents"""
+    if not del_request.document_ids:
+        raise errors.InvalidDocumentIdError("doc_ids can't be empty!")
+
+    for _id in del_request.document_ids:
+        validation.validate_id(_id)
+
+    if config.backend == enums.SearchDb.opensearch:
+        del_response: MqDeleteDocsResponse = delete_documents_marqo_os(config=config, deletion_instruction=del_request)
+    else:
+        raise RuntimeError(f"Config set to use unknown backend `{config.backend}`. "
+                           f"See tensor_search.enums.SearchDB for allowed backends")
+
+    return format_delete_docs_response(del_response)
+
+
+# -- Marqo-OS-specific deletion implementation: --
+
+
+def delete_documents_marqo_os(config: Config, deletion_instruction: MqDeleteDocsRequest) -> MqDeleteDocsResponse:
```

Comment:
Be careful with unbounded size on `deletion_instruction.document_ids`. There are limits to the size of HTTP requests. This is usually not a problem, but if you are sending millions of operations this can add up quickly.

Some other feedback.

- Consider chunking requests into, say, 10,000 operations and then sending these requests in parallel
- You should:
  - potentially attempt retries, assuming request is retryable based on response HTTP code (e.g. `429`)
  - manage back-offs
  - send multiple requests in parallel for higher throughput
- See [here](https://www.elastic.co/blog/indexing-documents-with-the-nest-elasticsearch-net-client) lookup "BulkAllObservable<T> helper" heading for inspiration
- Inspect the bulk response(s) to determine they have been successful
- Consider [disabling/changing the refresh interval](https://www.elastic.co/guide/en/elasticsearch/reference/current/tune-for-indexing-speed.html#_unset_or_increase_the_refresh_interval) around your bulk request (ensure it is correctly set back!)